File: Project/java/src/main/java/com/basanta/todo/Task.java
package com.basanta.todo;
import java.time.Instant;

/**
 * Task model: immutable object, validation.
 * Language-specific: final class, Instant for timestamps.
 */
public final class Task {
    private final String id;
    private final String title;
    private final String description;
    private final String category;
    private final String status;
    private final int priority;
    private final String assignee;
    private final String createdBy;
    private final String createdAt;
    private final String updatedAt;

    public Task(String id, String title, String description, String category, String status,
                int priority, String assignee, String createdBy, String createdAt, String updatedAt) {
        this.id = id;
        this.title = title;
        this.description = description != null ? description : "";
        this.category = category;
        this.status = status != null ? status : TaskStatus.PENDING;
        this.priority = priority;
        this.assignee = assignee != null ? assignee : "";
        this.createdBy = createdBy;
        this.createdAt = createdAt != null ? createdAt : Instant.now().toString();
        this.updatedAt = updatedAt != null ? updatedAt : this.createdAt;
    }

    public String getId() { return id; }
    public String getTitle() { return title; }
    public String getDescription() { return description; }
    public String getCategory() { return category; }
    public String getStatus() { return status; }
    public int getPriority() { return priority; }
    public String getAssignee() { return assignee; }
    public String getCreatedBy() { return createdBy; }
    public String getCreatedAt() { return createdAt; }
    public String getUpdatedAt() { return updatedAt; }

    public static Task create(String id, String title, String description, String category, String status,
                              int priority, String assignee, String createdBy) {
        String now = Instant.now().toString();
        return new Task(id, title, description, category, status != null ? status : TaskStatus.PENDING,
            priority, assignee, createdBy, now, now);
    }

    /** Immutable update: returns new Task (null means keep existing). */
    public Task withUpdates(String title, String description, String category, String status, Integer priority, String assignee) {
        return new Task(
            id,
            title != null ? title : this.title,
            description != null ? description : this.description,
            category != null ? category : this.category,
            status != null ? status : this.status,
            priority != null ? priority : this.priority,
            assignee != null ? assignee : this.assignee,
            createdBy,
            createdAt,
            Instant.now().toString()
        );
    }

    public static User.ValidationResult validateTitle(String title) {
        if (title == null || title.isBlank())
            return User.ValidationResult.error("Title is required");
        if (title.length() > 200)
            return User.ValidationResult.error("Title must be at most 200 characters");
        return User.ValidationResult.ok();
    }

    public static User.ValidationResult validateTaskInput(String title, String description, String category, String status, Integer priority) {
        User.ValidationResult tr = validateTitle(title);
        if (!tr.valid) return tr;
        if (!Category.isValidCategory(category))
            return User.ValidationResult.error("Invalid category. Must be one of: WORK, PERSONAL, SHOPPING, HEALTH, OTHER");
        if (status != null && !TaskStatus.isValidStatus(status))
            return User.ValidationResult.error("Invalid status. Must be one of: PENDING, IN_PROGRESS, COMPLETED");
        if (priority != null && (priority < 0 || priority > 10))
            return User.ValidationResult.error("Priority must be between 0 and 10");
        return User.ValidationResult.ok();
    }
}

File: Project/javaScript/src/Task.js
/**
 * Task module: task object, task methods, validation.
 */

import { isValidCategory } from './Category.js';
import { isValidStatus } from './TaskStatus.js';

/**
 * @typedef {Object} Task
 * @property {string} id
 * @property {string} title
 * @property {string} description
 * @property {string} category
 * @property {string} status
 * @property {number} priority
 * @property {string} assignee
 * @property {string} createdBy
 * @property {string} createdAt
 * @property {string} updatedAt
 */

/**
 * Create a task object.
 * @param {Object} opts
 * @param {string} opts.id
 * @param {string} opts.title
 * @param {string} [opts.description]
 * @param {string} opts.category
 * @param {string} [opts.status]
 * @param {number} [opts.priority]
 * @param {string} [opts.assignee]
 * @param {string} opts.createdBy
 * @returns {Task}
 */
export function createTask({
  id,
  title,
  description = '',
  category,
  status = 'PENDING',
  priority = 0,
  assignee = '',
  createdBy,
}) {
  const now = new Date().toISOString();
  return {
    id,
    title,
    description: description || '',
    category,
    status: status || 'PENDING',
    priority: typeof priority === 'number' ? priority : 0,
    assignee: assignee || '',
    createdBy,
    createdAt: now,
    updatedAt: now,
  };
}

/**
 * Update task fields (returns new task object).
 * @param {Task} task
 * @param {Partial<Pick<Task, 'title'|'description'|'category'|'status'|'priority'|'assignee'>>} updates
 * @returns {Task}
 */
export function updateTask(task, updates) {
  const updatedAt = new Date().toISOString();
  return {
    ...task,
    ...updates,
    updatedAt,
  };
}

/**
 * @param {string} title
 * @returns {{ valid: boolean, error?: string }}
 */
export function validateTitle(title) {
  if (typeof title !== 'string' || !title.trim()) {
    return { valid: false, error: 'Title is required' };
  }
  if (title.length > 200) {
    return { valid: false, error: 'Title must be at most 200 characters' };
  }
  return { valid: true };
}

/**
 * @param {Object} input
 * @param {string} input.title
 * @param {string} [input.description]
 * @param {string} input.category
 * @param {string} [input.status]
 * @param {number} [input.priority]
 * @returns {{ valid: boolean, error?: string }}
 */
export function validateTaskInput({ title, description, category, status, priority }) {
  const titleResult = validateTitle(title);
  if (!titleResult.valid) return titleResult;
  if (!isValidCategory(category)) {
    return { valid: false, error: `Invalid category. Must be one of: WORK, PERSONAL, SHOPPING, HEALTH, OTHER` };
  }
  if (status !== undefined && !isValidStatus(status)) {
    return { valid: false, error: `Invalid status. Must be one of: PENDING, IN_PROGRESS, COMPLETED` };
  }
  if (priority !== undefined && (typeof priority !== 'number' || priority < 0 || priority > 10)) {
    return { valid: false, error: 'Priority must be a number between 0 and 10' };
  }
  return { valid: true };
}

export default {
  createTask,
  updateTask,
  validateTitle,
  validateTaskInput,
};
